actors/market/src/emit.rs

```diff
@@ -0,0 +1,42 @@
+use fil_actors_runtime::runtime::Runtime;
+use fil_actors_runtime::{ActorError, EventBuilder};
+use fvm_shared::deal::DealID;
+use fvm_shared::ActorID;
+
+/// Indicates a deal has been published.
+pub fn deal_published(
+    rt: &impl Runtime,
+    client: ActorID,
+    provider: ActorID,
+    deal_id: DealID,
+) -> Result<(), ActorError> {
+    rt.emit_event(
+        &EventBuilder::new()
+            .typ("deal-published")
+            .field_indexed("client", &client)
+            .field_indexed("provider", &provider)
```

Comment:
@Stebalien for a second opinion, I suggested it might be worth including these fields since the deal ID is generated by the market actor, so a caller can't know what ID to filter for ahead of time. Many watchers will be interested in a small subset by party, and can avoid state inspection for most unwanted events with these fields.

However, by this logic we should also add client and provider to the verified allocation event too.

Reply:
1. Are we not including the deal label and/or CID? That seems like the more important identifier.
2. Regardless, including the provider/client definitely make sense here.
3. For the verified allocation event, I haven't thought enough about those to be sure? My guess is that the clients will want to track them (e.g., in a wallet), so including the client's ID is likely a good idea (at the very least).

Personally, I'd just include all fields we think might be relevant, then benchmark to make sure we haven't increased gas fees by too much.

Reply:
See https://github.com/filecoin-project/FIPs/discussions/754#discussioncomment-7204690 for why we're not including all the fields that might be relevant. Extending that discussion is still welcome.

Reply:
+1 to @Stebalien's idea of atleast including the client and provider to the events as well and then revisiting them after gas costs benchmarking. The UX will be much better if clients can reduce state lookups for events that have nothing to do with what they're interested in.

Reply:
Yeah, I guess I'm not exactly being consistent here (I think I was mostly focused on the discussion around identifying the event type).

I think we _do_ want to include the minimum amount of information necessary to:

1. Identify the event (so we can query for more information).
2. Determine if the event might be interesting (usually just the IDs of the parties involved). This may significantly decrease complexity/costs for developers.

At first blush, I wouldn't include things like amounts because they can be queried. However, I'm not entirely sure about that because queries:

1. Can only only see the state between top-level messages.
2. Can only _efficiently_ see the state between tipsets.

E.g., if a token is transferred multiple times within a single top-level message, being able to query the state isn't sufficient.

So, really, there's a third item: Any information the user might not be able to accurately and/or efficiently query for.

Reply:
Let's just do the parties involved for now. In the absence of events, tools like Sentinel have been able to calculate state changes by comparing before and after states.

Reply:
> tools like Sentinel have been able to calculate state changes by comparing before and after states.

I'd avoid relying on "sentinel managed to do it" as evidence of anything:

1. Sentinel mostly cares about the _state_ of the system, not the individual operations. People who work with tokens, etc., care about the individual transfers.
2. Sentinel actually had undecidable edge-cases (e.g., why did a sector terminate).
3. Sentinel relies on, e.g., the fact a user can't terminate and compact in a single top-level message. That kind of guarantee will eventually disappear as we allow more and more on-chain programmability.
4. Building and maintaining Sentinel was quite a bit more painful than it would have been if we had had proper events.

That doesn't mean we should just throw everything we can in every event (we _can_ always add more information later), just... avoid using sentinel as evidence.